data/transactions/logic/assembler.go

```diff
@@ -1184,17 +1150,177 @@ func typeLoads(pgm *ProgramKnowledge, args []string) (StackTypes, StackTypes) {
 	return nil, StackTypes{scratchType}
 }
 
-// keywords or "pseudo-ops" handle parsing and assembling special asm language
-// constructs like 'addr' We use an OpSpec here, but it's somewhat degenerate,
-// since they don't have opcodes or eval functions. But it does need a lot of
-// OpSpec, in order to support assembly - Mode, typing info, etc.
-var keywords = map[string]OpSpec{
-	"int":  {0, "int", nil, proto(":i"), 1, assembler(asmInt)},
-	"byte": {0, "byte", nil, proto(":b"), 1, assembler(asmByte)},
+func pseudoImmediatesError(ops *OpStream, name string, specs map[int]OpSpec) {
+	immediateCounts := make([]int, len(specs))
+	i := 0
+	for numImms := range specs {
+		immediateCounts[i] = numImms
+		i++
+	}
+	sort.SliceStable(immediateCounts, func(j, k int) bool { return immediateCounts[j] < immediateCounts[k] })
+	errMsg := fmt.Sprintf("%s expects", name)
+	for i, numImms := range immediateCounts {
+		if i+1 < len(immediateCounts) {
+			errMsg += fmt.Sprintf(" %d or", numImms)
+		} else {
+			errMsg += fmt.Sprintf(" %d", numImms)
+		}
+	}
+	errMsg += " immediate arguments"
+	ops.error(errMsg)
+}
+
+// getSpec finds the OpSpec we need during assembly based on it's name, our current version, and the immediates passed in
+// Note getSpec handles both normal OpSpecs and those supplied by versionedPseudoOps
+func getSpec(ops *OpStream, name string, args []string) (OpSpec, string, bool) {
+	pseudoSpecs, ok := ops.versionedPseudoOps[name]
+	if ok {
+		pseudo, ok := pseudoSpecs[len(args)]
+		if !ok {
+			// Could be that pseudoOp wants to handle immediates itself so check -1 key
+			pseudo, ok = pseudoSpecs[anyImmediates]
+			if !ok {
+				// Number of immediates supplied did not match any of the pseudoOps of the given name, so we try to construct a mock spec that can be used to track types
+				pseudoImmediatesError(ops, name, pseudoSpecs)
+				proto, version, ok := mergeProtos(pseudoSpecs)
+				if !ok {
+					return OpSpec{}, "", false
+				}
+				pseudo = OpSpec{Name: name, Proto: proto, Version: version, OpDetails: OpDetails{asm: func(*OpStream, *OpSpec, []string) error { return nil }}}
+			}
+		}
+		pseudo.Name = name
+		if pseudo.Version > ops.Version && (ops.Version != 0 || pseudo.Version != 1) {
+			switch {
+			case len(args) > 1:
+				ops.errorf("%s opcode with %d immediates was introduced in TEAL v%d", pseudo.Name, len(args), pseudo.Version)
```

Comment:
@jannotti Do you envision including the PR prior to _or_ after AVM 7 release?

If after, do you agree to rename away from _TEAL version_ towards _AVM version_ as per https://github.com/algorand/go-algorand/issues/4125?

```suggestion
				ops.errorf("%s opcode with %d immediates was introduced in AVM v%d", pseudo.Name, len(args), pseudo.Version)
```

Reply:
Following up on our verbal:  I'm on-board with dropping _TEAL_ here to leave only _version_.

Reply:
@iten-alg Resolving - Thanks for dropping _TEAL_ in _TEAL v_ references.